radae/dsp.py

```diff
@@ -51,28 +51,63 @@ def __init__(self, Ntap, Fs_Hz, bandwidth_Hz, centre_freq_Hz):
       # if true we can remove time flip from convolution
       assert np.all(self.h == np.flip(self.h))
 
-      self.mem = np.zeros(self.Ntap-1, dtype=np.csingle)
+      # Preallocate memory needed for BPF
+      self.mem = np.zeros(self.Ntap-1, dtype=np.complex64)
+      self.x_mem = np.zeros(self.Ntap + max_len - 1, dtype=np.complex64)
+      self.x_filt = np.zeros(max_len, dtype=np.complex64)
+      self.n = max_len
+
       self.phase = 1 + 0j
+      self.phase_vec_exp = np.exp(-1j*self.alpha*np.arange(1,max_len+1), dtype=np.complex64)
 
    def bpf(self, x):
       n = len(x)
-      phase_vec = self.phase*np.exp(-1j*self.alpha*np.arange(1,n+1))
-      x_baseband = x*phase_vec                                         # mix down to baseband
-      x_mem = np.concatenate([self.mem,x_baseband])                    # pre-pend filter memory
-      x_filt = np.zeros(n, dtype=np.csingle)
-      for i in np.arange(n):
-         x_filt[i] = np.dot(x_mem[i:i+self.Ntap],self.h)
-      self.mem = x_mem[-self.Ntap-1:]                                  # save filter state for next time
-      self.phase = phase_vec[-1]                                       # save phase state for next time
-      return x_filt*np.conj(phase_vec)                                 # mix back up to centre freq
+
+      # Make sure length is less than the filter length
+      assert(n <= self.n)
+
+      # Mix down to baseband
+      phase_vec = self.phase*self.phase_vec_exp[0:n]
+      x_baseband = x*phase_vec
+
+      # Store concatenated memory and baseband samples into x_mem
+      np.concatenate([self.mem, x_baseband], out=self.x_mem[0:len(x_baseband) + len(self.mem)])
+ 
```

Comment:
Just an observation - I guess this is kinda where Python gets awkward and we might as well be using C as we're having to think about in place versus pointers etc :thinking: Oh well, it is what it is.